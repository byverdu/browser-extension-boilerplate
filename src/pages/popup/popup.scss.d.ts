// *** This file is automatically generated - Do not edit! ***
/* tslint:disable */
/* eslint-disable */

declare const styles = {
    popup: 'popup__popup--1cPme',
    title: 'popup__title--3pMWf',
} as const;

export type ClassName = string;

export default styles;