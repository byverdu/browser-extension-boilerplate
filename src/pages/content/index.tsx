import React, { useState, useEffect, useRef } from 'react';
import { render } from 'react-dom';

import { wrapperBrowserAPI } from 'api';
import { Link,ExtensionMessages } from 'types';

import styles from './content.scss';

const { setStorage, sendMessage, getExtensionName } = wrapperBrowserAPI;
const extensionName = getExtensionName ();
const linksSaved: ExtensionMessages = 'links-saved'

const App = () => {
  const [ linksCount, setLinksCount ] = useState<number> ( 0 );
  const container = useRef<HTMLDivElement> ();
  const saveLinks = async ( links ) => {
    const oldStorage = await wrapperBrowserAPI.getStorage ( linksSaved );
    const newValues = {
      [ linksSaved ]: {
        ...oldStorage[ linksSaved ],
        ...links
      }
    };

    return await setStorage ( linksSaved, newValues );
  };

  useEffect ( () => {
    const links: Link[] = [
      ...document.querySelectorAll ( 'a' ),
    ]
      .map ( ( { href, textContent } ) => ( { href, textContent } ) )
      .filter ( ( link ) => link.textContent );

    saveLinks ( links ).then ( ( resp ) => {
      console.log ( 'storage', resp );
      sendMessage ( linksSaved )
        .then ( resp => {
          if ( typeof resp === 'boolean' ) {
            setLinksCount ( links.length );
          }
        } );
    } );
  } );

  useEffect ( () => {
    setTimeout ( () => {
      if ( container.current ) {
        container.current.remove ();
      }
    }, 5000 );
  } );

  return (
    <div ref={container} className={styles.content}>
      Hello.... this page has {linksCount} links on it
    </div>
  );
};

const root = document.createElement ( 'div' );
root.setAttribute ( 'id',  extensionName );
document.body.appendChild ( root );

render ( <App />, document.getElementById ( extensionName ) );
