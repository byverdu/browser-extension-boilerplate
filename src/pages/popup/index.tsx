import React, { useState, useEffect } from 'react';
import { render } from 'react-dom';

import { wrapperBrowserAPI } from 'api';
import { Link } from 'types';

import { List, Header } from 'Components';

const { getStorage, sendMessage } = wrapperBrowserAPI;

interface State{
  links: Link[]
  tabHostname: string
}

export const Popup = () => {
  const [ state, setState ] = useState<State> ( {
    links: [],
    tabHostname: '' 
  } );

  useEffect (  () => {
    sendMessage ( 'get-active-tab' ).then ( resp => {
      const tabHostname = new URL ( resp ).hostname;
      getStorage ( 'links-saved' ).then ( resp => {
        setState ( {
          links: resp[ 'links-saved' ] [ tabHostname ],
          tabHostname  
        } );
      } );
    } );
  }, [] );

  const { links, tabHostname } = state;

  return (
    <div>
      <Header title={tabHostname} />
      <List links={links} />
    </div>
  );
};

if ( process.env.ENV !== 'test' ) {
  render ( <Popup />, document.getElementById ( 'root' ) );
}
