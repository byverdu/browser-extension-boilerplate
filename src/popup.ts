import { wrapperBrowserAPI } from 'api';
import { ExtensionStorage } from './types';

const { getStorage } = wrapperBrowserAPI;

getStorage ( 'links-saved' ).then ( ( storage: ExtensionStorage ) => {
  const links = storage[ 'links-saved' ].map ( ( { textContent } ) => textContent ).join ( '-' );

  document.getElementById ( 'root' ).innerHTML = links;
} );