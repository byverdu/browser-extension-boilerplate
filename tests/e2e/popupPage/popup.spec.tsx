jest.mock ( '../../src/api' );
import React from 'react';
import { render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Popup } from '../../../src/pages/popup';

describe ( 'Popup page', () => {
  it ( 'should render a title with hostname', async () => {
    const { getByText } = render ( <Popup /> );
    await waitFor ( () => {
      expect ( getByText ( 'localhost' ) ).toBeInTheDocument ();
    } );
  } );

  it ( 'should render as many links as the content page had', async () => {
    const { container } = render ( <Popup /> );
    await waitFor ( () => {
      expect ( container.querySelectorAll ( 'li' ) ).toHaveLength ( 3 );
    } );
  } );

  it ( 'should match the snapshot', async () => {
    const { container } = render ( <Popup /> );

    await waitFor ( () => {
      expect ( container.firstChild ).toMatchSnapshot ( `<div>
        <h1>localhost</h1>
          <ul>
            <li><a href="http://localhost/first-link">first link</a></li>
            <li><a href="http://localhost/second-link">second link</a></li>
            <li><a href="http://localhost/third-link">third link</a></li>
          </ul>
        </div>` );
    } );
  } );
} );