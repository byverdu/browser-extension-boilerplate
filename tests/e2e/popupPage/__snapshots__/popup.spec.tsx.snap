// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup page should match the snapshot: <div>
        <h1>localhost</h1>
          <ul>
            <li><a href="http://localhost/first-link">first link</a></li>
            <li><a href="http://localhost/second-link">second link</a></li>
            <li><a href="http://localhost/third-link">third link</a></li>
          </ul>
        </div> 1`] = `
<div>
  <h1 />
  <ul />
</div>
`;
